gui/ThemeEngine.cpp

```diff
@@ -130,19 +130,19 @@ static const DrawDataInfo kDrawDataDefaults[] = {
 	{kDDWidgetBackgroundEditText,   "widget_textedit",  kDrawLayerBackground,   kDDNone},
 	{kDDWidgetBackgroundSlider,     "widget_slider",    kDrawLayerBackground,   kDDNone},
 
-	{kDDButtonIdle,                 "button_idle",      kDrawLayerBackground,   kDDNone},
```

Comment:
This change is IMO the most controversial one. :smile: 
Because you change here something that have been there for... almost 15 years!
And the code in place worked since then.

Drawing in the background has this benefit: you render once and every time something changes above it, it's easy to get back.
But it has this drawback: as soon as you need to change something about it you need a full redraw of the area and this is costly.
Drawing on foreground means that as soon as your widget changes, you need to fully render it.

In the case of a button, it means, with your change, that if you hover it, you render it fully once in foreground idle, a second time in foreground hovered and then, when the mouse leaves, a third time time in foreground idle.
Without your changes, it is rendered once in background idle at dialog creation, then, when hovering, rendered from background idle (fast), rendered fully in foreground hovered, and when the mouse leaves, rendered from background idle.
For this case, your change is not efficient at all.

When changing the button state, with your changes, only the button considered is rendered fully in foreground (using dialog background from buffer).
Without your changes, the full dialog area is redrawn from zero.
In this case your changes are clearly winning.

Now, one idea: what happens if you keep button idle as background, set up button idle as a parent of button idle and drawing in foreground?
This has the drawback of needing a render of button idle in background buffer and a render of disabled button in foreground afterwards but changing state should only affect the considered button.
On the other hand, hovering a button doesn't change any performance.

Reply:
My notes 100% agree with your analysis. However:
- please note that `kDDDropDownButtonDisabled` is already rendered in the foreground, so `kDDButtonDisabled` change is just matching that
- the double overdraw in the hover state (idle+hovered) has a negligible performance hit comparing to `g_gui.scheduleTopDialogRedraw` ... trust me, I'm the one testing it on 32 MHz. ;-) (but from a software architecture point of view it sucks, agreed)
- _if you keep button idle as background, set up button idle as a parent of button <ins>idle</ins> and drawing in foreground_ -- I guess you meant button **disabled** there? If so, it was a million dollar idea sir because it has achieved exactly what we wanted!

Reply:
Indeed I mixed up the names when writing it down. :smile: